intellij/src/saros/intellij/negotiation/ModuleConfigurationInitializer.java

```diff
@@ -0,0 +1,214 @@
+import java.util.concurrent.ConcurrentHashMap;
+import org.apache.log4j.Logger;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.jetbrains.jps.model.JpsElement;
+import org.jetbrains.jps.model.java.JavaResourceRootType;
+import org.jetbrains.jps.model.java.JavaSourceRootType;
+import org.jetbrains.jps.model.module.JpsModuleSourceRootType;
+import saros.filesystem.IProject;
+import saros.intellij.filesystem.IntelliJProjectImpl;
+import saros.repackaged.picocontainer.Startable;
+import saros.session.ISarosSession;
+import saros.session.ISessionListener;
+
+/**
+ * This class ensures that modules which were transferred as part of the project negotiation are
+ * correctly configured. This is done by loading the correct module configuration from the
+ * additional project negotiation data once the module is added to the session.
+ */
+public class ModuleConfigurationInitializer implements Startable {
+  private static final Logger log = Logger.getLogger(ModuleConfigurationInitializer.class);
+
+  private final ISarosSession session;
+
+  private final Map<Module, Map<String, String>> queuedModuleOptions;
+
+  private final ISessionListener sessionListener =
+      new ISessionListener() {
+
+        @Override
+        public void resourcesAdded(IProject project) {
+          applyModuleConfiguration(project);
+        }
+      };
+
+  public ModuleConfigurationInitializer(ISarosSession session) {
+    this.session = session;
+
+    this.queuedModuleOptions = new ConcurrentHashMap<>();
+  }
+
+  @Override
+  public void start() {
+    session.addListener(sessionListener);
+  }
+
+  @Override
+  public void stop() {
+    session.removeListener(sessionListener);
+  }
+
+  /**
+   * Enqueues module options for a module created as part of the project negotiation. This causes
+   * the passed options to be applied to the module once it is officially added to the session.
+   *
+   * @param module the module to apply the options to
+   * @param configurationOptions the configuration options to apply to the module
+   */
+  public void enqueueModuleConfigurationChange(
+      @NotNull Module module, @NotNull Map<String, String> configurationOptions) {
+
+    queuedModuleOptions.put(module, configurationOptions);
+  }
+
+  /**
+   * Updates the shared module with the queued module options. Does nothing if no such options are
+   * found in the held queue.
+   *
+   * @param wrappedModule the <code>IProject</code> representing the shared module
+   */
+  private void applyModuleConfiguration(@NotNull IProject wrappedModule) {
+    Module module = wrappedModule.adaptTo(IntelliJProjectImpl.class).getModule();
+
+    Map<String, String> configurationOptions = queuedModuleOptions.remove(module);
+
+    if (configurationOptions == null) {
+      log.debug("Skipped module root model adjustment as not options were found - " + module);
+
+      return;
+    }
+
+    ModuleRootModificationUtil.updateModel(
```

Comment:
Does this ensure that the update is called in the correct thread context ? As far as I remember projectAdded may be called from any thread.